### DISCUSSION TITLE
Scatter Plot 3D - different color for each point

### DISCUSSION THREAD

**Discussion by triuk** [alandefreitas/matplotplusplus#252] [Q&A]
Hi, I've found the matplot++ is very similar to matplotlib, but way faster.
I'd like to make scatter3d plot from XYZ coords with different mark sizes and different colors. I've found how to do most of it in your example plots, but I am still struggling with colors. There is an "color" example `scatter3(x, y, z, sizes, colors);`, but I am somehow missing the color pattern it generates when I change values, so I guess the fifth parameter is not connected to exact color, but gradient.
I'd like to do something like this (four model points):
<pre><code>std::vector< double> x = { 1, 2, 3, 4 };
std::vector< double> y = { 0, 0, 0, 0 };
std::vector< double> z = { 0, 0, 0, 0 };
std::vector< double> size = { 10, 20, 30, 40 };
std::vector< std::string> color = {"red", "magenta", "blue", "yellow"};
matplot::scatter3(x, y, z, size, color)->marker_face(true);
matplot::show();</code></pre>
Is it possible to put into `scatter3` the information about point's color? Or is there any good workaround?
Thank you in advance.

**Comment by alandefreitas** [comment]
> fifth parameter is not connected to exact color, but gradient

Exactly. They are numbers that get converted into the corresponding gradient in the colormap.

https://github.com/alandefreitas/matplotplusplus/blob/280659aa6835acdb16983e2af4fa704bca98a503/source/matplot/core/axes_type.h#L1136

To have exact colors, you would need to change the colormap to a colormap with the colors you want in sequence:

https://github.com/alandefreitas/matplotplusplus/blob/280659aa6835acdb16983e2af4fa704bca98a503/source/matplot/core/axes_type.h#L612

```cpp
auto ax = gca();
ax->colormap(/* ... */);
```

And then give `scatter3` the indexes you want for each point:

```cpp
std::vector<double> color = {0.0, 1.0, 2.0, 3.0}; // ...
matplot::scatter3(x, y, z, size, color)->marker_face(true);
```

**Comment by triuk** [comment]
I see, thank you for the quick response! So values from color vector are mapped on interpolated colormap.
I suppose it takes highest and lowest value from "color" and projects them to the highest and lowest values in the colormap(map).
My test code now looks like this:
<pre><code>
std::vector< std::vector< double>> map = {{1,0,0}, {0,0,1}, {0,1,0}, // RGB
                                          {0,1,1}, {1,0,1}, {1,1,0}};// CMY
matplot::colormap(map);
</pre></code>
My problem is, the {0,1,2,3} ; {1,2,3,4} ; {2,3,4,5} etc. all generate same colors = lowest value Red, highest Yellow, and other are interpolated in between. To be able to choose exact color by number, I must include eg 1 and 6 in the vector, like { 1, 1, 2, 6 } or { 1, 6, 2, 3 }.
So I need to be extremely careful, how I play it, because I wanted to mark points of interest in the 3Dscatter, which will change in time (visualize particle dynamic). So for example changing the last point's mark that was previously Yellow(6) in { 1, 1, 2, **6** } to Blue(3) { 1, 1, 2, **3** } will result in the same color and will instantly break Green(2) point if there is not any other Yellow(6) present.
Did I get it right? Is there a better solution, or do I have to deal with this somehow? :)

**Reply by alandefreitas** [reply]
Yes. Colormaps and scatter plots are used to represent ranges of values projected into the colormap. It all depends on what you want to achieve, but if you want to highlight individual values it's often better to have two scatter plots so you can have different logic for the common values and the highlighter values. For instance, you can use a range of colors for the regular values and some fixed color / size for the values you want to highlight.

**Comment by triuk** [comment]
Great idea to divide data into subsets. I've played with it a little and came with the colororder solution and hold() function suggested in examples, because setting colormap for the next data erased the previous for whole resulting plot and I've not figured out how to prevent it. So this is it, I am going to iterate through bunch of objects with XYZ-size attributes and I will divide them to separate plotting vectors according to desired color. Then merge it all with multiplot. I just must obey the plotting order (base->highlight->etc.) to get same colors every time. Is it good idea or would you suggest something even better?
Thank you!
```cpp
int main() {
    //auto [x, y, z, size] = generate_data();

    auto graph = matplot::gca();
    std::vector<std::string> newcolors = {"cyan", "magenta", "blue"};
    matplot::colororder(newcolors);  

    std::vector< double> x_basic = { 1, 2 };
    std::vector< double> y_basic = { 0, 0 };
    std::vector< double> z_basic = { 0, 0 };
    std::vector< double> size_basic = { 10, 20 };
    graph->scatter3(x_basic, y_basic, z_basic, size_basic)->marker_face(true);
    matplot::hold(matplot::on);

    std::vector< double> x_hl = { 3, 4, 5 };
    std::vector< double> y_hl = { 0, 0, 0 };
    std::vector< double> z_hl = { 0, 0, 0 };
    std::vector< double> size_hl = { 10, 20, 30 };
    graph->scatter3(x_hl, y_hl, z_hl, size_hl)->marker_face(true);
    matplot::hold(matplot::off);

    matplot::show();

    return 0;
}
```
EDIT:
I've just tried what will happen when there are not any data in one subset. The color is then applied to the nothing and it continues with the next one. That's great I just need to obey the plotting order :)
```cpp
int main() {
    auto [x_basic, y_basic, z_basic, size_basic, x_hl, y_hl, z_hl, size_hl] = collectPlotData();

    auto graph = matplot::gca();
    graph->colororder(std::vector<std::string> {"cyan", "magenta", "blue"});  
    
    graph->hold(matplot::on);
    graph->scatter3(x_basic, y_basic, z_basic, size_basic)->marker_face(true);
    graph->scatter3(std::vector<double> {}, std::vector<double> {}, std::vector<double> {}, std::vector<double> {})->marker_face(true);
    graph->scatter3(x_hl, y_hl, z_hl, size_hl)->marker_face(true);
    graph->hold(matplot::off);

    matplot::show();

    return 0;
}
```
I cleaned it up a little, seems like the line order of some functions does not matter.
